emit_IR.py

```diff
@@ -0,0 +1,243 @@
+            parent_block.add_inst(store)
+
+    def visit_Quote(self, quote: scheme.Quote) -> None:
+        if isinstance(quote.expr, scheme.SSym):
+            self.quoted = True
+            super().visit(quote.expr)
+            return
+
+        is_list = isinstance(quote.expr, scheme.SPair) and quote.expr.is_list
+        assert quote.expr is scheme.Nil or is_list
+
+        quoted_exprs = list(cast(scheme.SList, quote.expr))
+        nil_var = bytecode.Var(next(self.var_names))
+        nil_alloc = bytecode.AllocInst(
+            nil_var, bytecode.NumLit(scheme.SNum(0)))
+
+        cdr = nil_var
+        self.parent_block.add_inst(nil_alloc)
+
+        for expr in reversed(quoted_exprs):
+            pair_var = bytecode.Var(next(self.var_names))
+            pair_alloc = bytecode.AllocInst(
+                pair_var, bytecode.NumLit(scheme.SNum(2)))
+            self.parent_block.add_inst(pair_alloc)
+
+            expr_emitter = ExpressionEmitter(
+                self.parent_block, self.bb_names, self.var_names,
+                self.local_env, self.global_env, quoted=True)
+            expr_emitter.visit(expr)
+
+            store_car = bytecode.StoreInst(
+                pair_var, bytecode.NumLit(scheme.SNum(0)), expr_emitter.result)
+            self.parent_block.add_inst(store_car)
+
+            store_cdr = bytecode.StoreInst(
+                pair_var, bytecode.NumLit(scheme.SNum(1)), cdr
+            )
+            self.parent_block.add_inst(store_cdr)
+
+            cdr = pair_var
+
+        self.result = cdr
+
+    def visit_SCall(self, call: scheme.SCall) -> None:
+        assert not self.quoted, 'Non-primitives in quoted list unsupported'
+
+        func_expr_emitter = ExpressionEmitter(
+            self.parent_block, self.bb_names, self.var_names,
+            self.local_env, self.global_env)
+        func_expr_emitter.visit(call.func)
+
+        args: List[bytecode.Parameter] = []
+        arg_emitter: Optional[ExpressionEmitter] = None
+        self.end_block = func_expr_emitter.end_block
+        for arg in call.args:
+            arg_emitter = ExpressionEmitter(
+                self.end_block, self.bb_names, self.var_names,
+                self.local_env, self.global_env
+            )
+            arg_emitter.visit(arg)
+            args.append(arg_emitter.result)
+            self.end_block = arg_emitter.end_block
+
+        call_result_var = bytecode.Var(next(self.var_names))
+        call_instr = bytecode.CallInst(
+            call_result_var, func_expr_emitter.result, args)
+
+        if arg_emitter is None:
+            func_expr_emitter.end_block.add_inst(call_instr)
+        else:
+            arg_emitter.end_block.add_inst(call_instr)
+
+        self.result = call_result_var
+
+
+def name_generator(prefix: str) -> Iterator[str]:
+    count = 0
+    while True:
+        yield f'{prefix}{count}'
+        count += 1
+
+
+class EnvBindingNotFound(Exception):
```

Comment:
When I was reading custom exceptions stuff, I found things suggesting that you want all of your exceptions in one file, which is why we have the `errors.py`. I *think* this should go there?

Reply:
I'm familiar with two philosophies: having all the exceptions in one place, and having exceptions defined in the same module that they logically "originate" from. If we want to go with all the exceptions in errors.py that's fine with me. 

We should probably also have a custom base class for our custom exceptions, no?